Changes in module 'nf-core/cnvpytor/view'
--- modules/nf-core/cnvpytor/view/main.nf
+++ modules/nf-core/cnvpytor/view/main.nf
@@ -9,8 +9,6 @@
 
     input:
     tuple val(meta), path(pytor_files)
-    val bin_sizes
-    val output_format
 
     output:
     tuple val(meta), path("*.vcf"), emit: vcf      , optional: true
@@ -22,18 +20,23 @@
     task.ext.when == null || task.ext.when
 
     script:
-    def output_suffix = output_format ?: 'vcf'
-    def bins   = bin_sizes ?: '1000'
+    def output_suffix = task.ext.args?.output_format ?: 'vcf'
+    def bins   = task.ext.args?.bins ?: '1000'
+    def filter = task.ext.args?.filter ?: '' 
     def input  = pytor_files.join(" ")
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
 
     python3 <<CODE
-    import cnvpytor,os
+    import cnvpytor,os,json
     binsizes = "${bins}".split(" ")
     for binsize in binsizes:
         file_list = "${input}".split(" ")
-        app = cnvpytor.Viewer(file_list, params={} )
+        if filter != '':
+            filter_dict = json.loads(filter.replace("'inf'", "float('inf')"))
+        else:
+            filter_dict = {}
+        app = cnvpytor.Viewer(file_list, params=filter_dict )
         outputfile = "{}_{}.{}".format("${prefix}",binsize.strip(),"${output_suffix}")
         app.print_filename = outputfile
         app.bin_size = int(binsize)

************************************************************
